### ISSUE TITLE
Coin selection algorithm proposal

### ISSUE THREAD

**Issue by brunoerg** [bitcoin/bitcoin#23164]
I’ve been working on a coin selection algorithm using Evolutionary Algorithm. I am creating this issue to discuss the possibilities related to it. If here is not the best place to discuss it, feel free to close this issue. 

If you don’t know what an Evolutionary Algorithm is, here is a good definition: "EA is a subset of evolutionary computation, a generic population-based metaheuristic optimization algorithm. An EA uses mechanisms inspired by biological evolution, such as reproduction, mutation, recombination, and selection. Candidate solutions to the optimization problem play the role of individuals in a population, and the fitness function determines the quality of the solutions (see also loss function). Evolution of the population then takes place after the repeated application of the above operators". Anyway, I recommend you to study more about EA before trying to understand this proposal. 

In Evolutionary Algorithms we have genes, chromosomes, and population, for example:
![image](https://user-images.githubusercontent.com/19480819/135721326-4b8213c5-e9ad-4656-b6ef-f68dbbb5ddac.png)

So, to begin our algorithm, we must first create an initial population. The population will contain an arbitrary number of possible solutions to the problem, oftentimes called members. In this case, a gene is a UTXO, so, every chromosome is a set of UTXOs. 

```
member1 = [utxo1, utxo2, utxo3]
member2 = [uxto2, utxo5, utxo6, utxo7, utxo8]
...
```

To create the initial population, we can use the following approach (considering 5 members per population):

- 1 member composed of all UTXOs
- 1 member that selects randomly from the shuffled UTXOs until the target is exceeded
- 3 random members

Obs.: Most evolutionary algorithms set a length for the chromosome. However, for this approach, we don’t do it because we don't know how many UTXOs our final solution will use. 

Ok, having our initial population, it is time to evaluate each solution (fitness). To do it, we can use the waste metric, introduced in Bitcoin Core recently. So, our metric to evaluate the members is the cost of creating change, the excess selection amount, and the cost of spending inputs now as opposed to sometime in the future (when we expect to be able to consolidate inputs). See more: https://bitcoincore.reviews/22009

After evaluating each member, we can define what is the best one among them and build our next population (new generation).

Our new population will have (considering 5 members): 

- 3 members from mutation
- 1 member keeping the same chromosome of the best solution from the previous generation
- 1 new random member

Considering 50% mutation rate, we create 3 members copying the same chromosome of the best solution from the previous generation and applying mutation, like:

```
for (gene in chromosome) {
  const random_value = getRandomValueBetween0and1()
  if (random_value == 1) {
    gene = getRandomUtxo()
  }
}
```

Ok, now we have a new population, and then, we can repeat all the processes (fitness and mutation) N times (being N the number of generations), the best member of the last generation will be our final solution.

**Comment by lsilva01** [comment]
That sounds interesting. Could you provide a performance benchmark (or other relevant metrics) between the current implementation and the one suggested in this issue?

**Comment by brunoerg** [comment]
> That sounds interesting. Could you provide a performance benchmark (or other relevant metrics) between the current implementation and the one suggested in this issue?

Yeah! I'm doing it and as soon as I finish it I'll post here. I've implemented this algorithm in coinselect js (https://github.com/brunoerg/coinselect/blob/master/genetic.js) to be able to compare with other ones.

**Comment by brunoerg** [comment]
> It would also be interested in a comparison with repeated random draws. So when the EA does e.g. 10000 fitness evaluations, compare that with choosing the best individual from 10000 random draws. That should give some insight into how well the evolution itself actually works.

Interesting, but... "So when the EA does e.g. 10000 fitness evaluations, compare that with choosing the best individual from 10000 random draws" I have to evaluate that 10000 random draws to know what is the best one, didn't understand it..

**Comment by martinus** [comment]
@brunoerg sorry for not wording this clear... I've meant that it would be interesting to see if an optimization that uses EA to mutate & crossover is actually an improvement over just random draws. To make the comparison fair, the EA and random draws should do the same number of fitness evaluations. If the EA optimizes with 10000 evaluations to find the best individual, a fair comparison would be with an algorithm that simply randomly choose 10000 individuals to find the best. If EA performs consistently better, then this is a sign that the crossover & mutation actually is a benefit. If just randomly picking individuals perform better this is a sign that the crossover & mutation does not really work well to explore the fitness space. It could go both ways.

**Comment by brunoerg** [comment]
@martinus Cool! Going to test it and post the results here! Thank you!

**Comment by brunoerg** [comment]
@Xekyo Thanks, Murch! <br><br>


> I would imagine that a smaller mutation rate and a larger population would be helpful

Yeah, I'm gonna test with a smaller one, most algorithms use less than 1% for mutation rate. 

> Let's say the best solution A is composed of the three UTXOs {A1, A2, A3} and the second best solution B has the four UTXOs {B1, B2, B3, B4}. The UTXOs in each solution are sorted by value. The next generation would reuse A, and a few descendants generated by flipping a coin on whether to use A1 or B1, A2 or B2, A3 or B3, nothing or B4, with a small chance (5%?) of taking neither or both resulting e.g. in {A1, A2, B3} and {A1, B2, A3, B3}. Or perhaps by taking the start of A and the end of B and vice versa.

Interesting. It would be a great crossover and mutation.

> Then maybe generate two more such descendants from the second and third best

I think it would be interesting if we increase the population. Considering 5 members, the second/third best solutions might be really bad, we shouldn't consider them to create decendents.. However, if we consider 20 or 30 members, I'm sure the second or third best solutions are decent and we can use them to create the new ones.

If the best solution A is {A1, A2, A3} and the second best solution B is {B1, B2, B3, B4}, we will first apply crossover, right? So,  I think we can flip a coin  to decide the genes (50% A/50% B), in this case, it would generate some ones like:

A = {A1, A2, A3}
B = {B1, B2, B3, B4}
possible descendants:
X = {A1, A2, B3}
X = {A1, B2, B3}
X = {A1, A2, A3, B4}
X = {A1, B2, B3, B4}
...

And then, we can apply mutation, maybe we can do:
```
for gene in genes:
  random = getRandomValueBetween1And1000()
  if random == 1:
     //gets replaced by a random different one
  elif random == 2:
    //drop it

random = getRandomValueBetween1And1000()
if random > 1: 
  //push a new UTXO
```

Of course, it is just an example to show we can use what you suggested as mutation:
> have a small chance that a UTXO is dropped, gets replaced by a random different one, or a previously unselected UTXO is randomly added.

About the new generations, considering 20 members (A (the best one), B (second), C, D, E, F, G...) the decedents could be:
1 - A
2 - Crossover A/B + mutation
3 - Crossover A/B + mutation
4-  Crossover A/B + mutation
5 - Crossover B/C + mutation
6 - Crossover B/C + mutation
7 - Crossover B/C + mutation
8 - Crossover A/C + mutation
9 - Crossover A/C + mutation
10 - Crossover A/C + mutation
...
What do you think?

**Comment by Xekyo** [comment]
You could have multiple different ways of creating descendants. For the head-tail-swap, I was actually thinking like this:

AAA + BBBB = AABB + BBA // i.e. the head attaches to the tail of the other and vice versa to generate two new individuals while the cutting point wouldn't necessarily be the middle.

Another would pick either the A or B UTXO for each position:

AAA + BBBB = BABB

I assume that literature about EA would have additional suggestions, but I'm not sure whether the computational effort is defensible compared to more systematic approaches. Perhaps the calculation could be interrupted early if there hasn't been an improvement for ten generations beside the maximum generation count, but that wouldn't do anything for the worst case, of course.

**Comment by aureleoules** [comment]
Very interesting algorithm, have you continued working on it @brunoerg?

**Comment by brunoerg** [comment]
@aureleoules more or less, hope to do more simulations soon. Do you have any suggestions to improve the idea?

**Comment by aureleoules** [comment]
Not really sorry, but I'd love to see your draft.

**Comment by Xekyo** [comment]
I'm not sure whether I made myself clear above. I find evolutionary algorithms interesting, but I remain skeptical that they're a good approach for coin selection. I would suspect that they require a lot of computation and would be very slow, while I don't see why recombining different subsets of input selections would hone in on a better solution.

I would second what @martinus said above. Before getting too much into the details on this, it would be good to substantiate that it even produces useful results, e.g. by showing that the input sets improve over time as the EA runs, and that it can beat e.g. the best out of multiple random selections.

**Comment by priyanshiiit** [comment]
Related research paper: https://link.springer.com/article/10.1007/s40747-022-00799-2

**Comment by Xekyo** [comment]
Mh, their description of “How Bitcoin selects transaction inputs” matches the “Knapsack” algorithm, but BnB was merged to Bitcoin Core in 2018. They also cite my Master thesis as having been written in 2020 instead of 2016. Gonna have to give the whole paper a read some time.

**Comment by Xekyo** [comment]
I finished reading the paper. I did not find it insightful.
Their approach does not seem useful to me. It is optimizing for things that I would not consider important and is extremely inefficient in trying to attain them. The simulation results are implausible, making me think that their scenarios are not representative for the task at hand or their re-implementation of Bitcoin Core's coin selection not accurate.
